File: typescript/tour-of-restate/src/auxiliary/payment_client.ts
/*
 * Copyright (c) 2023 - Restate Software, Inc., Restate GmbH
 *
 * This file is part of the Tour of Restate Typescript handler API,
 * which is released under the MIT license.
 *
 * You can find a copy of the license in the file LICENSE
 * in the root directory of this repository or package or at
 * https://github.com/restatedev/tour-of-restate
 */

let i = 0;
export class PaymentClient {
  public static get() {
    return new PaymentClient();
  }
  async call(idempotencyKey: string, amount: number): Promise<boolean> {
    console.log(
      `Payment call succeeded for idempotency key ${idempotencyKey} and amount ${amount}`,
    );
    // do the call
    return true;
  }

  async failingCall(idempotencyKey: string, amount: number): Promise<boolean> {
    if (i >= 2) {
      console.log(
        `Payment call succeeded for idempotency key ${idempotencyKey} and amount ${amount}`,
      );
      i = 0;
      return true;
    } else {
      console.log(
        `Payment call failed for idempotency key ${idempotencyKey} and amount ${amount}. Retrying...`,
      );
      i = i + 1;
      throw new Error("Payment call failed");
    }
  }
}

File: patterns-use-cases/ticket-reservation/ticket-reservation-python/example/utils/payment_client.py
class PaymentClient:

    def __init__(self):
        self.i = 0

    async def call(self, idempotency_key: str, amount: float) -> bool:
        print(f"Payment call succeeded for idempotency key {idempotency_key} and amount {amount}")
        # do the call
        return True

    async def failing_call(self, idempotency_key: str, amount: float) -> bool:
        if self.i >= 2:
            print(f"Payment call succeeded for idempotency key {idempotency_key} and amount {amount}")
            i = 0
            return True
        else:
            print(f"Payment call failed for idempotency key {idempotency_key} and amount {amount}. Retrying...")
            self.i += 1
            raise Exception("Payment call failed")